### DISCUSSION TITLE
Global Variable not changing

### DISCUSSION THREAD

**Discussion by gepping** [jspsych/jsPsych#981] [Q&A] [answered]
Hello!  I have a Javascript issue with updating a global variable that I am using to break out of a for loop.  At the beginning of the code, I set:
`var move_onto_testing_phase = 0;`

so that when a portion correct on a block exceeds a minimum value, the value switches from 0 to 1 
```
var block_analysis = {
            type: 'html-keyboard-response',
            stimulus: function() {
                var completed_block = jsPsych.data.getLastTrialData().values()[0].block_number;
                var end_of_block_text = '<p>Complete with block '+String(completed_block)+',</p><p>Press any key to continue.</p>';
                var last_block_trials = jsPsych.data.get().filter({block_number: completed_block});
                var number_correct = last_block_trials.filter({accuracy: 1}).count();
                var block_accuracy = number_correct/trials;
                if (block_accuracy > min_accuracy) {
                    end_of_block_text = '<p>Nice work, you scored over '+(min_accuracy*100)+'% on this block.</p><p>Press any key to advance to the testing phase</p>';
                    move_onto_testing_phase = 1;
                } return end_of_block_text;
            }
        };
```
and I can, in turn, break out of the for loop that is causing my blocks to repeat
```
for (var i = 0; i < blocks; i++) {
            var block_introduction = {
                type: 'html-keyboard-response',
                stimulus: '<p>Ready to begin block '+String(i+1)+'?</p><p>If yes, press any key to continue.</p>'
            };
            timeline.push(block_introduction);
            timeline.push(learning_block_procedure);
            timeline.push(block_analysis);

            if (move_onto_testing_phase > 0.5) {
                break;
            }
        };
```
But, unfortunately, move_onto_testing_phase is not changing here when the accuracy minimum is exceeded.

Strangely, it does affect move_onto_testing_phase in another area of the code:
```
var learning_trial = {
            type: 'categorize-image',
            stimulus: jsPsych.timelineVariable('stimulus'),
            key_answer: jsPsych.timelineVariable('correct_response'),
            choices: ['s','k'],
            incorrect_text: 'Incorrect',
            prompt:'<p>Press S for category A and K for Category B</p>',
            show_stim_with_feedback: false,
            stimulus_duration: 5000,
            feedback_duration: 1000,
            trial_duration: 5000,
            data: {
                condition: jsPsych.timelineVariable('condition'),
                correct_response: jsPsych.timelineVariable('correct_response'),
            },
            on_finish: function(data) {
                var accuracy = 0;
                if (data.correct) {
                    accuracy = 1;
                }
                data.accuracy = accuracy;
                var trial_number = jsPsych.data.get().filter({trial_type: 'categorize-image'}).count();
                data.trial_number = trial_number;
                var block_number = Math.ceil(trial_number/(trials));
                data.block_number = block_number;
                data.status = move_onto_testing_phase
            },
        };
```
And I can tell it altered move_onto_testing_phase here because the number associated with status changes in my displayed data.

Long story short, why is my variable being changed in some parts of my code but not others?

**Comment by jodeleeuw** [comment, marked as answer]
Hi @gepping ,

This is a common conceptual issue with new jsPsych users. We're working on getting some better tutorial content to explain this.

> Long story short, why is my variable being changed in some parts of my code but not others?

The critical thing here is *when* the code is running. Your `for` loop is running before the experiment begins. All of the code that is constructing the timeline executes when the page loads. (If you stick a `console.log()` statement in the for loop you'll see when the break statement runs -- which I think is probably never in this case, because the `for` loop will end before the `if(...)` statement ever becomes `true`.)

The variable is correct in the data for the `learning_trial` object because the `on_finish` function doesn't execute until the trial is over. 

If you want to progress to the next part of the experiment conditionally, then a [looping timeline](https://www.jspsych.org/overview/timeline/#looping-timelines) should do the trick.

**Reply by gepping** [reply]
Hi Josh,
Thank you for your explanation and for pointing me towards the looping timeline!  Using the looping timeline made this much easier.
